File: next-frontend/api/definitions/models/release.ts
import { z } from "zod";

export const ReleaseUserSchema = z.object({
  avatar_url: z.string().nullable(),
  id: z.string(),
  username: z.string(),
});

export const ReleaseSchema = z.object({
  created_at: z.coerce.date(),
  id: z.string(),
  name: z.string(),
  released: z.boolean(),
  released_at: z.coerce.date().nullable(),
  released_by: ReleaseUserSchema.nullable(),
  seo_description: z.string().nullable(),
  seo_keywords: z.string().nullable(),
  seo_title: z.string().nullable(),
  short_description: z.string().nullish(),
  tags: z.array(z.string()),
  updated_at: z.coerce.date(),
  version: z.string(),
});

export const PublicReleaseSchema = z.object({
  name: z.string(),
  released_at: z.coerce.date(),
  released_by: ReleaseUserSchema.nullable(),
  short_description: z.string().nullish(),
  tags: z.array(z.string()),
  version: z.string(),
});

export const LatestUnseenReleaseSchema = z.object({
  release: PublicReleaseSchema,
  unseen: z.boolean(),
});

export type LatestUnseenRelease = z.infer<typeof LatestUnseenReleaseSchema>;
export type PublicRelease = z.infer<typeof PublicReleaseSchema>;
export type Release = z.infer<typeof ReleaseSchema>;
export type ReleaseUser = z.infer<typeof ReleaseUserSchema>;

File: backend/src/entity/release.rs
use chrono::{DateTime, Local};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ReleaseUser {
    pub id: String,
    pub username: String,
    pub avatar_url: Option<String>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Release {
    pub id: Uuid,
    pub version: String,
    pub name: String,
    pub short_description: Option<String>,
    pub released: bool,
    pub released_at: Option<DateTime<Local>>,
    pub released_by: Option<ReleaseUser>,
    pub tags: Vec<String>,
    pub seo_title: Option<String>,
    pub seo_description: Option<String>,
    pub seo_keywords: Option<String>,
    pub created_at: DateTime<Local>,
    pub updated_at: DateTime<Local>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct SeenRelease {
    pub release_id: Uuid,
    pub user_id: String,
    pub seen_at: DateTime<Local>,
}
